src/change_stream.ts

```diff
@@ -655,11 +629,22 @@ export class ChangeStream<
   hasNext(callback: Callback<boolean>): void;
   hasNext(callback?: Callback): Promise<boolean> | void {
     this._setIsIterator();
-    return maybePromise(callback, cb => {
-      this._getCursor((err, cursor) => {
-        if (err || !cursor) return cb(err); // failed to resume, raise an error
-        cursor.hasNext(cb);
-      });
+    return maybePromise(callback, callback => {
```

Comment:
Are we concerned at all about shadowing the `callback` variable here? and in the subsequent methods? This syntax makes it harder to understand what's actually going on

Reply:
I'm in favor of this, it's a good idea that we should always shadow "callback" inside maybePromise. The outer `callback` is nullish where as the inner one is always defined (either as a promise resolver or the user passed in callback). By shadowing the name we prevent code inside this scope from accidentally reaching up and calling a nullish variable. TS will also catch that but I think its confusing in the reverse, when reading "maybePromise" code I shouldn't have to see two callback variables in the scope, only one of them is ever the correct one to use so might as well use the same name again.

Reply:
This was intended - @nbbeeken feel free to chime in if I misinterpreted our conversation the other day.

The second argument to `maybePromise` is a function that `maybePromise` invokes.  `maybePromise` passes the callback argument in for us (which could be a Promise.resolve or something similar).  By shadowing the outer callback, we ensure that we will never acccidentally call the outer callback in the inner function.  For example:

```typescript
function foo(callback: Callback<any>) {
  return maybePromise(callback, cb => {
    callback('ooops, called the wrong callback');
  });
}
```

Reply:
I don't entirely buy that justification because these sorts of things should be protected against via regression testing, not code obfuscation; i.e., if it's a legitimate concern, then add a test to ensure that the outer callback isn't called. There's no way for someone just looking at this code to know whether the shadowing is intentional or not and whether we do or don't ever want to call that outer function.

Reply:
As a relevant aside: clever code, even when well-intentioned, tends to be difficult to maintain in the long term, since it almost inevitably forces other to play the "guess the intention" game

Reply:
I would actually prefer to add the https://eslint.org/docs/rules/no-shadow lint rule to avoid what is generally an antipattern due to readability issues. I think in this case, it's an instance of misdirected defensive programming. If it's not a case we're truly worried about, and it's not worth testing, then it's not worth coding for. If it's a case we are worried about, then we *should* test for it. People reading the code shouldn't have to guess why you decided to shadow a variable - whether it was purposeful or an accident, and what the purpose may have been.

Reply:
A summary from slack discussion with Neal: 
I personally favor good variable names and a no-shadow lint rule, but I don't think it needs to be in the scope of the change streams work; I do want a comment if we keep the shadowing because that's the very minimum you need for clarity of intent here, so, Bailey - you are welcome to either add distinct but descriptive callback names here, or add a comment to each instance of the shadowing with an explanation of why the shadowing is intentional

Reply:
I created a jira ticket, left a TODO and added a comment explaining the shadowing.